### PR TITLE
Create Feature #9 manage dashboard

### PR DESCRIPTION
**Pull request by tranvanduc10599**
Coding FE & API  for Dashboard Screen

### CODE REVIEW

File: src/PatientChecking/PatientChecking/Controllers/DashboardController.cs [hunk 1]
```diff
@@ -9,10 +11,37 @@ namespace PatientChecking.Controllers
...
             return View();
         }
+        [HttpGet("[controller]/getDashBoard")]
+        public async Task<IActionResult> GetDashBoardData()
+        {
+            int numOfPatients = await _patientService.GetNumberOfPatients();
+            int numOfPatientsInMonth = await _patientService.GetNumberOfPatientsInCurrentMonth();
+            int numOfAppointment = await _appointmentService.GetNumberOfAppointments();
+            int numOfAppointmentInMonth = await _appointmentService.GetNumberOfAppointmentsInCurrentMonth();
+            int numOfAppointmentInToday = await _appointmentService.GetNumberOfAppointmentsInToday();
```
Comment (minhphien) [inline src/PatientChecking/PatientChecking/Controllers/DashboardController.cs]: Please create a separate Model to combine all three values needed for the appointment. Those values must be proceeded in one sql command by writing only one service method named appointmentService.GetAppointmentSummary(), to reduce times transfer data between the application and sql server.
Apply the same for patient service.

File: src/PatientChecking/PatientChecking/Services/PatientService.cs [hunk 2]
```diff
@@ -8,5 +10,25 @@ namespace PatientChecking.Services
 {
     public class PatientService : IPatientService
     {
+        private readonly PatientCheckInContext patientCheckInContext;
```
Comment (minhphien) [inline src/PatientChecking/PatientChecking/Services/PatientService.cs]: coding convention: always use _  prefix for private fields

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml [hunk 3]
```diff
@@ -2,4 +2,67 @@
...
+                $(".total-appointment").text(data.numOfAppointments);
+                $(".appointment-inmonth").text(data.numOfAppointmentsInMonth + " new appointments/this month");
+                $(".appointment-intoday").text(data.numOfAppointmentsInToday + " new appointments/today");
+                $('#loading-div-background').hide();
+                $(".grid-wide").show();
+            }
+        })
+    })();
+</script>
+<div id="loading-div-background" style="position: fixed; top: 50%; left: 60%;">
```
Comment (minhphien) [inline src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: please don't include in-line styling, instead try to use bootstrap class.

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml [hunk 4]
```diff
@@ -2,4 +2,67 @@
...
+    <div id="loading-div" class="ui-corner-all">
+        <div id="loadingbar"><i class="fa fa-spinner fa-spin fa-3x fa-fw"></i></div>
+        <h2 style="color:gray;font-weight:normal;">Please wait....</h2>
```
Comment (minhphien) [inline src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: see commend above

**Comment by minhphien** [comment]
@tranvanduc10599 @harryul  there are example applying bootstrap example here: https://getbootstrap.com/docs/5.1/examples/
Besides referecing Bootstrap documentation, I recommend you guys usually inspect their code and apply what they use to our application.

File: src/PatientChecking/PatientChecking/Controllers/DashboardController.cs [hunk 5]
```diff
@@ -9,10 +11,37 @@ namespace PatientChecking.Controllers
...
+        {
+            _patientService = patientService;
+            _appointmentService = appointmentService;
+        }
+
         public IActionResult Home()
         {
             return View();
         }
+        [HttpGet("[controller]/getDashBoard")]
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/Controllers/DashboardController.cs:28]: Add a new line above this method

File: src/PatientChecking/PatientChecking/Controllers/DashboardController.cs (hunk 1 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Controllers/DashboardController.cs]: I created separate models for patients and appointments to combine all values needed for each

**Review by vietvo361 — APPROVED** [review]

File: src/PatientChecking/PatientChecking/Services/PatientService.cs (hunk 2 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Services/PatientService.cs]: "coding convention: always use _ prefix for private fields" has been applied

File: src/PatientChecking/PatientChecking/Services/AppointmentService.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+        public async Task<int> GetNumberOfAppointmentsInCurrentMonth()
+        {
+            var result = await patientCheckInContext.Appointments.Where(x => x.CheckInDate.Month == DateTime.Now.Month).ToListAsync();
+            return result.Count;
+        }
+
+        public async Task<int> GetNumberOfAppointmentsInToday()
+        {
+            var result = await patientCheckInContext.Appointments.Where(x => x.CheckInDate.Date.Day == DateTime.Now.Day && x.Status == "CheckIn").ToListAsync();
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/Services/AppointmentService.cs]: Should move "CheckIn" to Appointment status enum.

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml [hunk 7]
```diff
@@ -2,4 +2,67 @@
...
+    <link href="@Url.Content("~/css/Dashboard.css")" rel="stylesheet" type="text/css" />
+}
+<script>
+    (function () {
+        $('#loading-div-background').show();
+        
+        $.ajax({
+            type: 'GET',
+            url: 'https://localhost:5001/Dashboard/getDashBoard',
+            success: function (data) {
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml:15]: We should handle error case. Alert user to reload the page and try again.

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml [hunk 8]
```diff
@@ -2,4 +2,67 @@
...
+@section css {
+    <link href="@Url.Content("~/css/Dashboard.css")" rel="stylesheet" type="text/css" />
+}
+<script>
+    (function () {
+        $('#loading-div-background').show();
+        
+        $.ajax({
+            type: 'GET',
+            url: 'https://localhost:5001/Dashboard/getDashBoard',
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: Should use: @Url.Action("getDashBoard", "Dashboard")

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml [hunk 9]
```diff
@@ -2,4 +2,67 @@
     ViewData["Title"] = "Dashboard";
     ViewData["Page"] = "Dashboard";
 }
-<p>This is dashboard</p>
+@section css {
+    <link href="@Url.Content("~/css/Dashboard.css")" rel="stylesheet" type="text/css" />
+}
+<script>
+    (function () {
+        $('#loading-div-background').show();
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: Can we move this loading to common file due to this file can use to other places?

File: src/PatientChecking/PatientChecking/Views/ViewModels/DashboardViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+namespace PatientChecking.Views.ViewModels
+{
+    public class DashboardViewModel
+    {
+        public int NumOfPatients { get; set; }
+        public int NumOfPatientsInMonth { get; set; }
+        public int NumOfAppointments { get; set; }
+        public int NumOfAppointmentsInMonth { get; set; }
+        public int NumOfAppointmentsInToday { get; set; }
+
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/Views/ViewModels/DashboardViewModel.cs]: Remove this empty line

File: src/PatientChecking/PatientChecking/wwwroot/css/Dashboard.css [hunk 11]
```diff
@@ -0,0 +1,116 @@
+﻿.grid-wide{
+    display:none;
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/wwwroot/css/Dashboard.css]: Add a space after ":"

File: src/PatientChecking/PatientChecking/wwwroot/css/Dashboard.css [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    z-index: 99;
+    opacity: 0.8;
+    filter: alpha(opacity=80);
+    -moz-opacity: 0.8;
+    min-height: 100%;
+    width: 100%;
+}
+
+.loading {
+    font-family: Arial;
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/wwwroot/css/Dashboard.css]: Remove this "font-family" for consistent with the whole page

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml [hunk 13]
```diff
@@ -2,4 +2,67 @@
...
+                <div class="right-item-patient">
+                    <p class="total-patient"></p>
+                    <p class="patient-inmonth"></p>
+                </div>
+            </div>
+        </div>
+        <div class="col">
+            <div class="item-appointment">
+                <div class="left-item-patient">
+                    <h3>
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml:53]: This is `<h3>` and end with `</h2>` :D ?

**Comment by vietvo361** [comment]
Please fix all comments. I mistakenly pressed the Approve button

**Review by vietvo361 — CHANGES_REQUESTED** [review]
Please fix all comments. I mistakenly pressed the Approve button @tranvanduc10599

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml (hunk 7 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml:15]: added case error when calling api

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml (hunk 8 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: Changed url: 'https://localhost:5001/Dashboard/getDashBoard', => url: '@Url.Action("getDashBoard", "Dashboard")',

File: src/PatientChecking/PatientChecking/Views/ViewModels/DashboardViewModel.cs (hunk 10 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Views/ViewModels/DashboardViewModel.cs]: I created separate models for patients and appointments to combine all values needed for each, so i removed DashboardViewModel

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml (hunk 9 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: I customized Loading tag, so we can use to other places by import:  <link href="@Url.Content("~/css/Loading.css")" rel="stylesheet" type="text/css" /> to each razor view

File: src/PatientChecking/PatientChecking/Services/PatientService.cs (hunk 2 above)
Comment (minhphien) [reply src/PatientChecking/PatientChecking/Services/PatientService.cs]: I don't see the fix being applied ?

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml [hunk 14]
```diff
@@ -2,4 +2,70 @@
...
+                $(".total-patient").text(data.patient.numOfPatients);
+                $(".patient-inmonth").text(data.patient.numOfPatientsInMonth + " new paitents/this month");
+                $(".total-appointment").text(data.appointment.numOfAppointments);
+                $(".appointment-inmonth").text(data.appointment.numOfAppointmentsInMonth + " new appointments/this month");
+                $(".appointment-intoday").text(data.appointment.numOfAppointmentsInToday + " new appointments/today");
+                $('.loader').hide();
+                $(".grid-wide").show();
+            }, error: function (jqXHR, textStatus, errorThrown) {
+                console.log('AJAX call failed.');
+                console.log(textStatus + ': ' + errorThrown);
```
Comment (minhphien) [inline src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: Please do not include console.log()

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml (hunk 3 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: I customized Loading tag, so we can use to other places

File: src/PatientChecking/PatientChecking/Services/AppointmentService.cs (hunk 6 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Services/AppointmentService.cs]: I created enum for Appointment Status

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml (hunk 14 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: I removed all commands console.log()

File: src/PatientChecking/PatientChecking/Services/ServiceModels/Enum.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace PatientChecking.Services.ServiceModels
+{
+    public class Enum
```
Comment (vietvo361) [inline src/PatientChecking/PatientChecking/Services/ServiceModels/Enum.cs]: We don't need `Enum` class, just:
```
public enum AppointmentStatus { 
  CheckIn = 1, 
  Closed = 2, 
  Cancel = 3 
}
```
Please rename `StatusAppointment` to `AppointmentStatus`

File: src/PatientChecking/PatientChecking/Controllers/DashboardController.cs (hunk 5 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Controllers/DashboardController.cs:28]: I added a new line above this method

File: src/PatientChecking/PatientChecking/wwwroot/css/Dashboard.css (hunk 11 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/wwwroot/css/Dashboard.css]: I added a space after ":"

File: src/PatientChecking/PatientChecking/wwwroot/css/Dashboard.css (hunk 12 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/wwwroot/css/Dashboard.css]: I removed font-family: Arial;

File: src/PatientChecking/PatientChecking/Services/ServiceModels/Enum.cs (hunk 15 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Services/ServiceModels/Enum.cs]: I separated Appointment Status into enum folder

File: src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml (hunk 4 above)
Comment (tranvanduc10599) [reply src/PatientChecking/PatientChecking/Views/Dashboard/Home.cshtml]: I customized Loading tag, so we can use to other places

**Pull request merged**